libbpf-tools/tcppktlat.c

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2023 Wenbo Zhang
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <time.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <sys/resource.h>
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include "tcppktlat.h"
+#include "tcppktlat.skel.h"
+#include "compat.h"
+#include "trace_helpers.h"
+
+static struct env {
+	pid_t pid;
+	pid_t tid;
+	__u64 min_us;
+	__u16 lport;
+	__u16 rport;
+	bool timestamp;
+	bool verbose;
+} env = {};
+
+static volatile sig_atomic_t exiting = 0;
+static int column_width = 15;
+
+const char *argp_program_version = "tcppktlat 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Trace latency between TCP received pkt and picked up by userspace thread.\n"
+"\n"
+"USAGE: tcppkglat [--help] [-T] [-p PID] [-t TID] [-l LPORT] [-r RPORT] [-v]\n"
+"\n"
+"EXAMPLES:\n"
+"    tcppkglat             # Trace all TCP packet picked up latency\n"
+"    tcppkglat -T          # summarize with timestamps\n"
+"    tcppkglat -p          # filter for pid\n"
+"    tcppkglat -t          # filter for tid\n"
+"    tcppkglat -l          # filter for local port\n"
+"    tcppkglat -r          # filter for remote port\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process PID to trace"},
+	{ "tid", 't', "TID", 0, "Thread TID to trace"},
+	{ "timestamp", 'T', NULL, 0, "include timestamp on output" },
+	{ "lport", 'l', "LPORT", 0, "filter for local port" },
+	{ "rport", 'r', "RPORT", 0, "filter for remote port" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+	long long min_us;
+	int pid;
+
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case 'p':
+		errno = 0;
+		pid = strtol(arg, NULL, 10);
+		if (errno || pid <= 0) {
+			fprintf(stderr, "Invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		env.pid = pid;
```

Comment:
In the current libbpf-tools, there is no problem with this place, but when I re-implement the acquisition of pid, I will check whether the pid exists in the system. through the interface:

```
static inline bool do_process_running(int pid)                                  
{                                                                               
        bool ret = kill(pid, 0);                                                
                                                                                
        if (ret)                                                                
                warning("PID %d is not running.\n", pid);                       
                                                                                
        return !ret;                                                            
}  
```

But this is not a big problem, and I see that other implementations basically don't pay attention to this, Never mind this, I'm just making a suggestion.